core/storage/hamt/hamt.cpp

```diff
@@ -0,0 +1,247 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "storage/hamt/hamt.hpp"
+
+#include "codec/cbor/cbor.hpp"
+#include "common/which.hpp"
+#include "crypto/blake2/blake2b160.hpp"
+#include "crypto/murmur/murmur.hpp"
+
+OUTCOME_CPP_DEFINE_CATEGORY(fc::storage::hamt, HamtError, e) {
+  using fc::storage::hamt::HamtError;
+  switch (e) {
+    case HamtError::EXPECTED_CID:
+      return "Expected CID";
+    case HamtError::NOT_FOUND:
+      return "Not found";
+    case HamtError::MAX_DEPTH:
+      return "Max depth exceeded";
+    default:
+      return "Unknown error";
+  }
+}
+
+namespace fc::storage::hamt {
+  using fc::common::which;
+
+  const CID kDummyCid({}, {}, libp2p::multi::Multihash::create({}, {}).value());
+
+  // assuming 8-bit indices
+  auto keyToIndices(const std::string &key, int n = -1) {
+    std::vector<uint8_t> key_bytes;
+    for (auto c : key) {
+      key_bytes.push_back(c);
+    }
+    auto hash = crypto::murmur::hash(key_bytes);
+    std::vector<size_t> result;
+    for (auto i = n == -1 ? hash.begin() : hash.end() - n + 1; i != hash.end();
```

Comment:
Please take the initialization of i outside with an ordinary if. The way it is now is hard to read.